chainfactory.go

```diff
@@ -108,107 +53,84 @@ var builtinChainConfigs = map[string]ibc.ChainConfig{
 }
 
 // NewBuiltinChainFactory returns a BuiltinChainFactory that returns chains defined by entries.
-func NewBuiltinChainFactory(entries []BuiltinChainFactoryEntry, logger *zap.Logger) *BuiltinChainFactory {
-	return &BuiltinChainFactory{entries: entries, log: logger}
+func NewBuiltinChainFactory(log *zap.Logger, specs []*ChainSpec) *BuiltinChainFactory {
+	return &BuiltinChainFactory{log: log, specs: specs}
 }
 
 func (f *BuiltinChainFactory) Count() int {
-	return len(f.entries)
+	return len(f.specs)
 }
 
 func (f *BuiltinChainFactory) Chains(testName string) ([]ibc.Chain, error) {
-	chains := make([]ibc.Chain, len(f.entries))
-	for i, e := range f.entries {
-		chain, err := e.GetChain(f.log, testName)
+	chains := make([]ibc.Chain, len(f.specs))
+	for i, s := range f.specs {
+		cfg, err := s.Config()
+		if err != nil {
+			// Prefer to wrap the error with the chain name if possible.
+			if s.Name != "" {
+				return nil, fmt.Errorf("failed to build chain config %s: %w", s.Name, err)
+			}
+
+			return nil, fmt.Errorf("failed to build chain config at index %d: %w", i, err)
+		}
+
+		chain, err := buildChain(f.log, testName, *cfg, s.NumValidators, s.NumFullNodes)
 		if err != nil {
 			return nil, err
 		}
 		chains[i] = chain
 	}
+
 	return chains, nil
 }
 
-func (f *BuiltinChainFactory) Name() string {
-	parts := make([]string, len(f.entries))
-	for i, e := range f.entries {
-		parts[i] = e.Name + "@" + e.Version
+func buildChain(log *zap.Logger, testName string, cfg ibc.ChainConfig, numValidators, numFullNodes *int) (ibc.Chain, error) {
+	nv := 2
```

Comment:
maybe this and `nf` could be defined slightly further up in the file as something like `defaultNumValidators`?